File: apps/api/src/main/java/com/parkflow/modules/parking/operation/validation/PlateValidator.java
package com.parkflow.modules.parking.operation.validation;

import java.util.List;
import java.util.Locale;
import java.util.Optional;
import java.util.ArrayList;
import java.util.regex.Pattern;
import org.springframework.stereotype.Component;

@Component
public class PlateValidator {

    private final List<PlateValidationRule> rules = List.of(
        new PlateValidationRule("CO", "CAR", Pattern.compile("^[A-Z]{3}[0-9]{3}$"), "ABC123", "Para carro en Colombia se esperan 3 letras y 3 números", true),
        new PlateValidationRule("CO", "VAN", Pattern.compile("^[A-Z]{3}[0-9]{3}$"), "ABC123", "Para van en Colombia se esperan 3 letras y 3 números", true),
        new PlateValidationRule("CO", "TRUCK", Pattern.compile("^[A-Z]{3}[0-9]{3}$"), "ABC123", "Para camión en Colombia se esperan 3 letras y 3 números", true),
        new PlateValidationRule("CO", "BUS", Pattern.compile("^[A-Z]{3}[0-9]{3}$"), "ABC123", "Para bus en Colombia se esperan 3 letras y 3 números", true),
        new PlateValidationRule("CO", "ELECTRIC", Pattern.compile("^[A-Z]{3}[0-9]{3}$"), "ABC123", "Para eléctrico en Colombia se esperan 3 letras y 3 números", true),
        new PlateValidationRule("CO", "MOTORCYCLE", Pattern.compile("^[A-Z]{3}[0-9]{2}[A-Z]{1}$"), "ABC12A", "Para moto en Colombia se esperan 3 letras, 2 números y 1 letra", true),
        new PlateValidationRule("CO", "BICYCLE", Pattern.compile("^[A-Z0-9]{3,12}$"), "BICI001", "Para bicicleta use un identificador de 3 a 12 letras o números", true),
        new PlateValidationRule("CO", "OTHER", Pattern.compile("^[A-Z]{3}[0-9]{3}$"), "ABC123", "Para vehículo general en Colombia se esperan 3 letras y 3 números", true)
    );

    /**
     * Normalizes a plate by converting to uppercase and removing spaces, dashes, and special characters.
     */
    public String normalizePlate(String plate) {
        if (plate == null) {
            return "";
        }
        return plate.toUpperCase(Locale.ROOT).replaceAll("[^A-Z0-9]", "");
    }

    /**
     * Validates a plate according to country and vehicle type rules.
     */
    public PlateValidationResult validatePlate(String countryCode, String vehicleType, String plate) {
        String normalized = normalizePlate(plate);
        
        if (normalized.isEmpty()) {
            return PlateValidationResult.invalid(normalized, "La placa no puede estar vacía o contener solo caracteres especiales");
        }

        // Default to CO if not provided for now, or you could make it strict
        String targetCountry = (countryCode != null && !countryCode.isBlank()) ? countryCode : "CO";
        String targetType = (vehicleType != null && !vehicleType.isBlank()) ? vehicleType : "OTHER";

        Optional<PlateValidationRule> activeRule = rules.stream()
            .filter(r -> r.enabled() && r.countryCode().equalsIgnoreCase(targetCountry) && r.vehicleType().equalsIgnoreCase(targetType))
            .findFirst();

        if (activeRule.isEmpty()) {
            boolean isCountrySupported = rules.stream().anyMatch(r -> r.countryCode().equalsIgnoreCase(targetCountry));
            if (isCountrySupported) {
                 return PlateValidationResult.invalid(normalized, "Tipo de vehículo no soportado para validación en " + targetCountry);
            }
            return PlateValidationResult.invalid(normalized, "Pais no soportado para validacion de placa: " + targetCountry);
        }

        PlateValidationRule rule = activeRule.get();
        if (rule.pattern().matcher(normalized).matches()) {
            return PlateValidationResult.valid(normalized);
        }

        // It didn't match. Let's see if it matched ANOTHER vehicle type's rule to give a better error message.
        Optional<PlateValidationRule> crossMatch = rules.stream()
            .filter(r -> r.enabled() && r.countryCode().equalsIgnoreCase(targetCountry) && !r.vehicleType().equalsIgnoreCase(targetType))
            .filter(r -> r.pattern().matcher(normalized).matches())
            .findFirst();

        String errorMessage = rule.errorMessage() + " (Ej: " + rule.example() + ").";
        if (crossMatch.isPresent()) {
            errorMessage += " Parece que ingresaste una placa de " + translateVehicleType(crossMatch.get().vehicleType()) + ".";
        }

        return PlateValidationResult.invalid(normalized, errorMessage);
    }

    /**
     * Infers the most likely vehicle type from plate format for a given country.
     * Returns empty if no rule matches.
     */
    public Optional<String> inferVehicleType(String countryCode, String plate) {
        String normalized = normalizePlate(plate);
        if (normalized.isEmpty()) {
            return Optional.empty();
        }

        String targetCountry = (countryCode != null && !countryCode.isBlank()) ? countryCode : "CO";
        List<String> matches = new ArrayList<>();

        for (PlateValidationRule rule : rules) {
            if (!rule.enabled()) continue;
            if (!rule.countryCode().equalsIgnoreCase(targetCountry)) continue;
            if (rule.pattern().matcher(normalized).matches()) {
                if (!matches.contains(rule.vehicleType().toUpperCase(Locale.ROOT))) {
                    matches.add(rule.vehicleType().toUpperCase(Locale.ROOT));
                }
            }
        }

        if (matches.isEmpty()) return Optional.empty();
        if (matches.contains("MOTORCYCLE")) return Optional.of("MOTORCYCLE");
        if (matches.contains("CAR")) return Optional.of("CAR");
        return Optional.of(matches.get(0));
    }

    private String translateVehicleType(String type) {
        return switch (type.toUpperCase(Locale.ROOT)) {
            case "CAR" -> "carro";
            case "MOTORCYCLE" -> "moto";
            case "VAN" -> "van";
            case "TRUCK" -> "camión";
            case "BUS" -> "bus";
            case "BICYCLE" -> "bicicleta";
            case "ELECTRIC" -> "eléctrico";
            default -> "otro vehículo";
        };
    }
}

File: apps/web/src/lib/validation/plate-validator.ts
export interface PlateValidationRule {
  countryCode: string;
  vehicleType: string;
  pattern: RegExp;
  example: string;
  errorMessage: string;
  enabled: boolean;
}

export const plateRules: PlateValidationRule[] = [
  { countryCode: "CO", vehicleType: "CAR", pattern: /^[A-Z]{3}[0-9]{3}$/, example: "ABC123", errorMessage: "Para carro en Colombia se esperan 3 letras y 3 números", enabled: true },
  { countryCode: "CO", vehicleType: "VAN", pattern: /^[A-Z]{3}[0-9]{3}$/, example: "ABC123", errorMessage: "Para van en Colombia se esperan 3 letras y 3 números", enabled: true },
  { countryCode: "CO", vehicleType: "TRUCK", pattern: /^[A-Z]{3}[0-9]{3}$/, example: "ABC123", errorMessage: "Para camión en Colombia se esperan 3 letras y 3 números", enabled: true },
  { countryCode: "CO", vehicleType: "BUS", pattern: /^[A-Z]{3}[0-9]{3}$/, example: "ABC123", errorMessage: "Para bus en Colombia se esperan 3 letras y 3 números", enabled: true },
  { countryCode: "CO", vehicleType: "ELECTRIC", pattern: /^[A-Z]{3}[0-9]{3}$/, example: "ABC123", errorMessage: "Para eléctrico en Colombia se esperan 3 letras y 3 números", enabled: true },
  { countryCode: "CO", vehicleType: "MOTORCYCLE", pattern: /^[A-Z]{3}[0-9]{2}[A-Z]{1}$/, example: "ABC12A", errorMessage: "Para moto en Colombia se esperan 3 letras, 2 números y 1 letra", enabled: true },
  { countryCode: "CO", vehicleType: "BICYCLE", pattern: /^[A-Z0-9]{3,12}$/, example: "BICI001", errorMessage: "Para bicicleta use un identificador de 3 a 12 letras o números", enabled: true },
  { countryCode: "CO", vehicleType: "OTHER", pattern: /^[A-Z0-9]{2,10}$/, example: "CD1234", errorMessage: "Para vehículo general se esperan entre 2 y 10 caracteres alfanuméricos", enabled: true }
];

export function normalizePlate(plate: string): string {
  if (!plate) return "";
  return plate.toUpperCase().replace(/[^A-Z0-9]/g, "");
}

export interface PlateValidationResult {
  isValid: boolean;
  normalizedPlate: string;
  errorMessage?: string;
}

export function inferVehicleType(countryCode: string | undefined | null, plate: string): string | null {
  const normalized = normalizePlate(plate);
  if (!normalized) return null;

  const targetCountry = (countryCode || "CO").toUpperCase();
  const matches = plateRules
    .filter((r) => r.enabled && r.countryCode.toUpperCase() === targetCountry && r.pattern.test(normalized))
    .map((r) => r.vehicleType.toUpperCase());

  if (matches.length === 0) return null;
  if (matches.includes("MOTORCYCLE")) return "MOTORCYCLE";
  if (matches.includes("CAR")) return "CAR";
  return matches[0] ?? null;
}

export const translateVehicleType = (type: string) => {
  switch (type.toUpperCase()) {
    case 'CAR': return 'carro';
    case 'MOTORCYCLE': return 'moto';
    case 'VAN': return 'van';
    case 'TRUCK': return 'camión';
    case 'BUS': return 'bus';
    case 'BICYCLE': return 'bicicleta';
    case 'ELECTRIC': return 'eléctrico';
    default: return 'otro vehículo';
  }
};

export function getPlateFormatHint(vehicleType: string, countryCode = "CO"): string {
  const targetCountry = countryCode.toUpperCase();
  const targetType = vehicleType.toUpperCase();
  const rule = plateRules.find(r => r.enabled && r.countryCode.toUpperCase() === targetCountry && r.vehicleType.toUpperCase() === targetType);
  if (!rule) return "3 letras + 3 números";
  const parts = rule.errorMessage.replace(/^Para\s+\S+\s+en\s+\S+\s+se\s+esperan\s+/i, "").replace(/\.$/, "");
  return parts || "3 letras + 3 números";
}

export function getPlatePlaceholder(vehicleType: string, countryCode = "CO"): string {
  const targetCountry = countryCode.toUpperCase();
  const targetType = vehicleType.toUpperCase();
  const rule = plateRules.find(r => r.enabled && r.countryCode.toUpperCase() === targetCountry && r.vehicleType.toUpperCase() === targetType);
  return rule?.example || "ABC123";
}

export function validatePlate(countryCode: string | undefined | null, vehicleType: string, plate: string): PlateValidationResult {
  const normalized = normalizePlate(plate);
  
  if (!normalized) {
    return { isValid: false, normalizedPlate: normalized, errorMessage: "La placa no puede estar vacía o contener solo símbolos" };
  }

  const targetCountry = countryCode || "CO";
  const targetType = vehicleType || "OTHER";

  const activeRule = plateRules.find(r => r.enabled && r.countryCode.toUpperCase() === targetCountry && r.vehicleType.toUpperCase() === targetType);

  if (!activeRule) {
    const isCountrySupported = plateRules.some(r => r.countryCode.toUpperCase() === targetCountry);
    if (isCountrySupported) {
      return { isValid: false, normalizedPlate: normalized, errorMessage: `Tipo de vehículo no soportado para validación en ${targetCountry}` };
    }
    return { isValid: false, normalizedPlate: normalized, errorMessage: `Pais no soportado para validación de placa: ${targetCountry}` };
  }

  if (activeRule.pattern.test(normalized)) {
    return { isValid: true, normalizedPlate: normalized };
  }

  const crossMatch = plateRules.find(r => r.enabled && r.countryCode.toUpperCase() === targetCountry && r.vehicleType.toUpperCase() !== targetType && r.pattern.test(normalized));

  let errorMessage = `${activeRule.errorMessage} (Ej: ${activeRule.example}).`;
  if (crossMatch) {
    errorMessage += ` Parece que ingresaste una placa de ${translateVehicleType(crossMatch.vehicleType)}.`;
  }

  return { isValid: false, normalizedPlate: normalized, errorMessage };
}
